File: leetcode/3Sum.cpp
//============================================================================
// Name        :
// Author      : hussam
// Version     :
// Copyright   :
// Description :https://leetcode.com/problems/3sum/description/
//============================================================================

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>>res;
        sort(nums.begin(),nums.end());
        int se,la;
        for(int fi=0;fi<nums.size();fi++)
        
        {
            if(fi==0||nums[fi]!=nums[fi-1])
            {
            se=fi+1;
            la=nums.size()-1;
            while(se<la)
            {
            if(nums[fi]+nums[se]+nums[la]==0)
            {
                vector<int>tmp(3);
                tmp[0]=nums[fi];
                tmp[1]=nums[se];
                tmp[2]=nums[la];
                res.push_back(tmp);
           
           while(se<la&&nums[se]==nums[se+1])
           se++;
           while(se<la&&nums[la]==nums[la-1])
           la--;
           
           se++;
           la--;
               
            }
            else if(nums[fi]+nums[se]+nums[la]<0)
            {
                se++;
               
            }
           else
           la--;
          
            }
            
            }
           
            
                
            }
            
        
        return res;
    }
};

File: leetcode/two pointers/meeting/3sum.py
# Author : Hussam
# description : https://leetcode.com/problems/3sum/description/
# solution : two pointers
# o(n^2) time, o(n) space


class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # if array is empty
        if nums is None or len(nums) < 3:
            return []
        # sort numbers and implement two pointer
        result = []
        nums.sort()
        for i in range(len(nums)):
            if i == 0 or nums[i] != nums[i - 1]:
                left, right = i + 1, len(nums)-1
                while left < right:
                    if nums[i] + nums[left] + nums[right] == 0:
                        result.append((nums[i], nums[left], nums[right]))
                        while left < right and nums[left] == nums[left + 1]:
                            left += 1
                        while left < right and nums[right] == nums[right - 1]:
                            right -= 1
                        left += 1
                        right -= 1
                    elif nums[i] + nums[left] + nums[right] < 0:
                        left += 1
                    else:
                        right -= 1
        return result
